sync/optimistic.md

```diff
@@ -0,0 +1,290 @@
+
+A consensus engine MUST be able to retrospectively (i.e., after import) modify
+the status of `SYNCING` blocks to be either `VALID` or `INVALID` based upon responses
+from an execution engine. I.e., perform the following transitions:
+
+- `SYNCING` -> `VALID`
+- `SYNCING` -> `INVALID`
+
+When a block transitions from `SYNCING` -> `VALID`, all *ancestors* of the
+block MUST also transition from `SYNCING` -> `VALID`. Such a block is no longer
+considered "optimistically imported".
+
+When a block transitions from `SYNCING` -> `INVALID`, all *descendants* of the
+block MUST also transition from `SYNCING` -> `INVALID`.
+
+When a block transitions from the `SYNCING` state it is removed from the set of
+`optimistic_roots`.
+
+### Execution Engine Errors
+
+When an execution engine returns an error or fails to respond to a payload
+validity request for some block, a consensus engine:
+
+- MUST NOT optimistically import the block.
+- MUST NOT apply the block to the fork choice store.
+- MAY queue the block for later processing.
+
+### Assumptions about Execution Engine Behaviour
+
+This specification assumes execution engines will only return `SYNCING` when
+there is insufficient information available to make a `VALID` or `INVALID`
+determination on the given `ExecutionPayload` (e.g., the parent payload is
+unknown). Specifically, `SYNCING` responses should be fork-specific, in that
+the search for a block on one chain MUST NOT trigger a `SYNCING` response for
+another chain.
+
+### Re-Orgs
+
+The consensus engine MUST support any chain reorganisation which does *not*
+affect the justified checkpoint. The consensus engine MAY support re-orgs
+beyond the justified checkpoint.
+
+If the justified checkpoint transitions from `SYNCING` -> `INVALID`, a
+consensus engine MAY choose to alert the user and force the application to
+exit.
+
+## Fork Choice
+
+Consensus engines MUST support removing blocks from fork choice that transition
+from `SYNCING` to `INVALID`. Specifically, a block deemed `INVALID` at any
+point MUST NOT be included in the canonical chain and the weights from those
+`INVALID` blocks MUST NOT be applied to any `VALID` or `SYNCING` ancestors.
+
+### Fork Choice Poisoning
+
+During the merge transition it is possible for an attacker to craft a
+`BeaconBlock` with an execution payload that references an
+eternally-unavailable `body.execution_payload.parent_hash` (i.e., the parent
+hash is random bytes). In rare circumstances, it is possible that an attacker
+can build atop such a block to trigger justification. If an optimistic node
+imports this malicious chain, that node will have a "poisoned" fork choice
+store, such that the node is unable to produce a block that descends from the
+head (due to the invalid chain of payloads) and the node is unable to produce a
+block that forks around the head (due to the justification of the malicious
+chain).
+
+If honest chain exists which justifies a higher epoch than the malicious chain,
```

Comment:
It should be enough for a node to import an honest chain justifying an epoch *prior to* malicious chain that justifies the *same* epoch to avoid poisoning the fork choice store and eventually finishing an optimistic sync which in this case should converge on the honest chain

Reply:
>  an honest chain justifying an epoch prior to malicious chain that justifies the same epoch 

I'm not sure I follow this sorry.

Reply:
> If an honest chain exists which justifies a higher epoch than the malicious chain, that chain will take precedence and revive any poisoned store.

Do we need a *higher* epoch to be justified by an honest chain? From my perspective the *same* epoch justified by an honest chain is enough to prevent poisoned store, with the condition that an honest justification is applied to the store *before* the malicious one. Simply:

1) A node imports malicious chain that justifies epoch `N`
2) The store is poisoned and rejects blocks of an honest chain that has not yet justified `N`

VS

1) A node imports an honest chain that justifies epoch `N`
2) A node imports a malicious chain that justifies epoch `N`
3) Despite of (2) the store is healthy and is capable of keep applying blocks of an honest chain

An honest chain justifying `N+1` does also prevent store poisoning. My point is that justifying `N` is also enough. And this is what we assume to get in 4 epochs: an honest chain with adversary power `< 1/3` justifies epoch `N`. Then we allow for optimistically import *any* other chain without the risk of a store being poisoned.

Am I missing anything?

Reply:
>  with the condition that an honest justification is applied to the store before the malicious one

Ah yes, I see what you mean now. So a higher epoch *always* fixes it and a lower one fixes it *sometimes*, given a message ordering assumption.

I've mentioned this "poisoning prevention" in 6d73b0a4 :)